yamux/src/frame/header.rs

```diff
@@ -323,65 +348,47 @@ impl fmt::Display for StreamId {
 impl nohash_hasher::IsEnabled for StreamId {}
 
 /// Possible flags set on a message.
-#[derive(Copy, Clone, Debug, PartialEq, Eq)]
-pub struct Flags(u16);
+#[derive(Copy, Clone, Debug, PartialEq, Eq, FromBytes, AsBytes, FromZeroes)]
+#[repr(packed)]
+pub struct Flags(U16);
 
 impl Flags {
     pub fn contains(self, other: Flags) -> bool {
-        self.0 & other.0 == other.0
+        self.0.get() & other.0.get() == other.0.get()
     }
 
     pub fn val(self) -> u16 {
-        self.0
+        self.0.get()
     }
 }
 
 /// Indicates the start of a new stream.
-pub const SYN: Flags = Flags(1);
+pub const SYN: Flags = Flags(U16::from_bytes([0, 1]));
 
 /// Acknowledges the start of a new stream.
-pub const ACK: Flags = Flags(2);
+pub const ACK: Flags = Flags(U16::from_bytes([0, 2]));
 
 /// Indicates the half-closing of a stream.
-pub const FIN: Flags = Flags(4);
+pub const FIN: Flags = Flags(U16::from_bytes([0, 4]));
 
 /// Indicates an immediate stream reset.
-pub const RST: Flags = Flags(8);
+pub const RST: Flags = Flags(U16::from_bytes([0, 8]));
 
 /// The serialised header size in bytes.
 pub const HEADER_SIZE: usize = 12;
 
-/// Encode a [`Header`] value.
-pub fn encode<T>(hdr: &Header<T>) -> [u8; HEADER_SIZE] {
-    let mut buf = [0; HEADER_SIZE];
-    buf[0] = hdr.version.0;
-    buf[1] = hdr.tag as u8;
-    buf[2..4].copy_from_slice(&hdr.flags.0.to_be_bytes());
-    buf[4..8].copy_from_slice(&hdr.stream_id.0.to_be_bytes());
-    buf[8..HEADER_SIZE].copy_from_slice(&hdr.length.0.to_be_bytes());
-    buf
-}
-
-/// Decode a [`Header`] value.
+// Decode a [`Header`] value.
 pub fn decode(buf: &[u8; HEADER_SIZE]) -> Result<Header<()>, HeaderDecodeError> {
```

Comment:
I'm not convinced that the decoding should be done on the raw buffer, as we have introduced zerocopy and can work instead on the struct header.

Reply:
It isn't really good use of the type system to allow construction of types that are in an invalid state. Thus, until `zerocopy` can give us these guarantees, I'd rather have a dedicated `decode` function. This somewhat ensures that we are not working with a `Header` that has invalid values set.